docs/enhancements/on-prem-data-collection.md

```diff
@@ -0,0 +1,121 @@
+When a cluster installation attempt finishes (regardless of success or failure), all of the event data for this cluster will be queried from the DB, packaged, and sent to the data hosting endpoint.
+
+Changes that need to be made in the `Assisted-Service` to support this:
+- Needs to know what the data hosting service endpoint is and how to authenticate into it
+- Needs code changes when an install is attempted to query data
+- Needs to know how to format the data to send to the data hosting service
+- Needs to expose an option for users to opt out of this feature
+
+Additional changes needed:
+- Documentation regarding terms and conditions?
+- Documentation about how to opt out
+- Documentation of requirements: non-disconnected environment
+
+Sending data should not interfere with customer workloads and deployment
+
+### User Stories
+
+#### Story 1
+
+As an infrastructure operator, I can provision and manage my fleet of clusters knowing that data points are automatically collected and sent to the Assisted Installer team without having to do additional configurations. I know my data will be safe and secure and the team will only collect data they need to improve the product. 
+
+
+#### Story 2
+
+As a developer on the assisted installer team, I can analyze the customer data to determine if a feature is worth implementing/keeping/improving. I know that the customer data is accurate and up-to-date. All of the data is parse-able and can be easily tailored to the graphs/visualizations that help my analysis.
+
+#### Story 3
+
+As a product owner, I can determine if the product is moving in the right direction based on the actual customer data. I can help determine which features and bug fixes to prioritize based on the data.
+
+### Implementation Details/Notes/Constraints
+
+Assisted Service changes:
+1. New environment variables to be added:
+    - `DISABLE_DATA_COLLECTION`: If set to `true`, it will disable this feature and thereby allow a customer to opt-out, otherwise the feature is on by default.
+    - `DATA_COLLECTION_URL`: The URL the data will be sent to.
+2. Add additional logic when a cluster reaches one of the following end stages: `error`, `cancelled`, `installed`
+    - Each of these cluster stages represent an end state and that the cluster attempted installation.
+    - Additional logic includes: querying the DB for events of that cluster and streaming this event data to the `DATA_COLLECTION_URL` over HTTPS.
+        - If the customer has large-scale clusters, would it be better to always compress the data in a separate thread?
+3. Add logic for authenticating to the `DATA_COLLECTION_URL` 
+    - Currently unknown since the service to host this data hasn't been decided on yet, but it could take the form of using a Kubernetes Secret
+
+### Risks and Mitigations
+
+On-premises users might have large-scale clusters so network usage and processor usage could be impacted depending on how large the data set is. 
```

Comment:
what will happen if the on prem environment will continue to send the data over and over again and will cause to overload in the data flow? we need to have the ability to stop it

Reply:
Good point! @rccrdpccl or @tsorya do you happen to know if there's a way to check for duplicate data in either the ingress (pipelines) or the kafka? Like if the cluster id & user email already exists?

---

If not then either a task to investigate this or an alternative solution needs to be added. 

---
An option for the alternative solution could be to include a new data field for each event to show whether or not it's been sent to streaming. The field could also only be filled if it's an on-prem install.

A problem could be that the field marks the data as sent but the connection or packets get dropped somewhere and never make it to the external service.

Reply:
We can handle duplication from the data pipeline (we should expose event_id though, that currently it isn't).
The main issue in my opinion is avoiding DDoS the http insights ingress service

Reply:
Oh I see, thanks for clarifying that!

Some ideas to prevent DDoS-ing is maybe:

1. Only sending the events when the cluster has reached installation or failure to install stages
2. Having a max retry for sending the events to the insights ingress service
3. Not retrying until a timeout has been reached and there's no response from the insights ingress service

What do you think? Are there better ways to prevent this?

Reply:
I think they are good points (point 1 is also discussed [here](https://github.com/openshift/assisted-service/pull/4457#discussion_r985767135)).
I would also consider retrying with exponential backoff and jitter, to prevent all clients retrying at the same time in case of downtime of our service and using remote configuration for these settings and possibly the service URL and circuit breaking, this way we have some control over client applications (i.e. possibility to configure `DATA_COLLECTION_URL` through a remote manifest  `https://example.com/data_collection.json` that the application would read first - this can be cached in a CDN or something so it can be retrieved many times). The advantage of doing this is that 1) we can send traffic to /dev/null 2) if we still want to try retrieve the data we'd have a few knobs to facilitate it.

Reply:
Great ideas! Agreed with exponential backoff and jitter.

For the remote manifest, how often should the assisted service try to read it? I would think:
1. When the assisted service initially starts
2. Every time the assisted service will send data it'll get this configuration first

If the manifest is unreachable, then we can default to the last configuration we got or the default if it was never able to reach it.

> The advantage of doing this is that 1) we can send traffic to /dev/null

By this, do you mean we can prevent DDoS-ing the http insights ingress service by directing `DATA_COLLECTION_URL` to `/dev/null`?

Reply:
> For the remote manifest, how often should the assisted service try to read it? I would think:
> 
> 1. When the assisted service initially starts
> 2. Every time the assisted service will send data it'll get this configuration first
>
> If the manifest is unreachable, then we can default to the last configuration we got or the default if it was never able to reach it.

Sounds good to me (both when to read it and behaviour when it's unreachable)

> By this, do you mean we can prevent DDoS-ing the http insights ingress service by directing `DATA_COLLECTION_URL` to `/dev/null`?

Yes, really by `/dev/null` I mean any URL that wouldn't overload the main one.
A secondary effect is that if we have a change in the ingress URL for any reason, we can update the URL without breaking older clients.

Reply:
I've added the remote manifest and retrying as a future enhancement